File: apps/web-app/src/lib/validation/schemas/account.ts
import { z } from 'zod';
import { createWalletSchema } from './wallet';

/**
 * Account type enum
 */
export const accountTypeSchema = z.enum(['admin', 'merchant']);

/**
 * Schema for creating an account
 */
export const createAccountSchema = z.object({
  name: z.string()
    .min(1, 'Name is required')
    .max(100, 'Name must be less than 100 characters'),
  account_type: accountTypeSchema,
  description: z.string()
    .max(500, 'Description must be less than 500 characters')
    .optional(),
  business_name: z.string()
    .min(1)
    .max(100, 'Business name must be less than 100 characters')
    .optional(),
  business_type: z.string()
    .max(50, 'Business type must be less than 50 characters')
    .optional(),
  website_url: z.string()
    .url('Invalid website URL')
    .optional()
    .or(z.literal('')),
  support_email: z.string()
    .email('Invalid email format')
    .optional(),
  support_phone: z.string()
    .regex(/^\+?[1-9]\d{1,14}$/, 'Invalid phone number format')
    .optional(),
  finished_onboarding: z.boolean().optional(),
  metadata: z.record(z.unknown()).optional(),
  // Web3Auth embedded wallet data
  wallet_data: createWalletSchema.optional(),
});

/**
 * Schema for account onboarding request
 */
export const accountOnboardingSchema = z.object({
  address_line1: z.string()
    .min(1)
    .max(100, 'Address line 1 must be less than 100 characters')
    .optional(),
  address_line2: z.string()
    .max(100, 'Address line 2 must be less than 100 characters')
    .optional(),
  city: z.string()
    .min(1)
    .max(50, 'City must be less than 50 characters')
    .optional(),
  state: z.string()
    .length(2, 'State must be 2 characters')
    .optional(),
  postal_code: z.string()
    .regex(/^\d{5}(-\d{4})?$/, 'Invalid postal code format')
    .optional(),
  country: z.string()
    .length(2, 'Country must be 2-letter ISO code')
    .optional(),
  first_name: z.string()
    .min(1)
    .max(50, 'First name must be less than 50 characters')
    .optional(),
  last_name: z.string()
    .min(1)
    .max(50, 'Last name must be less than 50 characters')
    .optional(),
  wallet_address: z.string()
    .regex(/^0x[a-fA-F0-9]{40}$/, 'Invalid Ethereum address format')
    .optional(),
  finished_onboarding: z.boolean().optional(),
});

/**
 * Schema for updating an account
 */
export const updateAccountSchema = z.object({
  name: z.string()
    .min(1)
    .max(100, 'Name must be less than 100 characters')
    .optional(),
  description: z.string()
    .max(500, 'Description must be less than 500 characters')
    .optional(),
  business_name: z.string()
    .min(1)
    .max(100, 'Business name must be less than 100 characters')
    .optional(),
  business_type: z.string()
    .max(50, 'Business type must be less than 50 characters')
    .optional(),
  website_url: z.string()
    .url('Invalid website URL')
    .optional()
    .or(z.literal('')),
  support_email: z.string()
    .email('Invalid email format')
    .optional(),
  support_phone: z.string()
    .regex(/^\+?[1-9]\d{1,14}$/, 'Invalid phone number format')
    .optional(),
  account_type: accountTypeSchema.optional(),
  finished_onboarding: z.boolean().optional(),
  metadata: z.record(z.unknown()).optional(),
});

/**
 * Schema for account ID parameter
 */
export const accountIdParamSchema = z.object({
  accountId: z.string().uuid('Invalid account ID format'),
});

// Type exports
export type CreateAccountInput = z.infer<typeof createAccountSchema>;
export type AccountOnboardingInput = z.infer<typeof accountOnboardingSchema>;
export type UpdateAccountInput = z.infer<typeof updateAccountSchema>;
export type AccountType = z.infer<typeof accountTypeSchema>;

File: libs/go/types/api/requests/account.go
package requests

// CreateAccountRequest represents the request body for creating an account
type CreateAccountRequest struct {
	Name               string                 `json:"name" binding:"required"`
	AccountType        string                 `json:"account_type" binding:"required,oneof=admin merchant"`
	Description        string                 `json:"description,omitempty"`
	BusinessName       string                 `json:"business_name,omitempty"`
	BusinessType       string                 `json:"business_type,omitempty"`
	WebsiteURL         string                 `json:"website_url,omitempty"`
	SupportEmail       string                 `json:"support_email,omitempty"`
	SupportPhone       string                 `json:"support_phone,omitempty"`
	FinishedOnboarding bool                   `json:"finished_onboarding,omitempty"`
	Metadata           map[string]interface{} `json:"metadata,omitempty"`
	// Web3Auth embedded wallet data to be created during registration
	WalletData *EmbeddedWalletRequest `json:"wallet_data,omitempty"`
}

// EmbeddedWalletRequest represents the request body for creating an embedded wallet during account creation
type EmbeddedWalletRequest struct {
	WalletType    string                 `json:"wallet_type" binding:"required"` // 'web3auth', 'wallet', or 'circle_wallet'
	WalletAddress string                 `json:"wallet_address" binding:"required"`
	NetworkType   string                 `json:"network_type" binding:"required"`
	Nickname      string                 `json:"nickname,omitempty"`
	ENS           string                 `json:"ens,omitempty"`
	IsPrimary     bool                   `json:"is_primary"`
	Verified      bool                   `json:"verified"`
	Metadata      map[string]interface{} `json:"metadata,omitempty"`
	// Circle wallet specific fields
	CircleUserID   string `json:"circle_user_id,omitempty"`   // Only for circle wallets
	CircleWalletID string `json:"circle_wallet_id,omitempty"` // Only for circle wallets
	ChainID        int32  `json:"chain_id,omitempty"`         // Only for circle wallets
	State          string `json:"state,omitempty"`            // Only for circle wallets
}

// UpdateAccountRequest represents the request body for updating an account
type UpdateAccountRequest struct {
	Name               string                 `json:"name,omitempty"`
	Description        string                 `json:"description,omitempty"`
	BusinessName       string                 `json:"business_name,omitempty"`
	BusinessType       string                 `json:"business_type,omitempty"`
	WebsiteURL         string                 `json:"website_url,omitempty"`
	SupportEmail       string                 `json:"support_email,omitempty"`
	SupportPhone       string                 `json:"support_phone,omitempty"`
	AccountType        string                 `json:"account_type,omitempty" binding:"omitempty,oneof=admin merchant"`
	FinishedOnboarding bool                   `json:"finished_onboarding,omitempty"`
	Metadata           map[string]interface{} `json:"metadata,omitempty"`
}

// OnboardAccountRequest represents the request body for onboarding an account
type OnboardAccountRequest struct {
	FirstName     string `json:"first_name,omitempty"`
	LastName      string `json:"last_name,omitempty"`
	AddressLine1  string `json:"address_line1"`
	AddressLine2  string `json:"address_line2"`
	City          string `json:"city"`
	State         string `json:"state"`
	PostalCode    string `json:"postal_code"`
	Country       string `json:"country"`
	TaxID         string `json:"tax_id,omitempty"`
	BusinessPhone string `json:"business_phone,omitempty"`
	WalletAddress string `json:"wallet_address,omitempty"`
}
